mergify_engine/worker_lua.py

```diff
@@ -0,0 +1,212 @@
+# -*- encoding: utf-8 -*-
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import datetime
+import hashlib
+import typing
+import uuid
+
+import aredis.exceptions
+import daiquiri
+
+from mergify_engine import github_types
+from mergify_engine import utils
+
+
+LOG = daiquiri.getLogger(__name__)
+
+ScriptIdT = typing.NewType("ScriptIdT", uuid.UUID)
+
+SCRIPTS: typing.Dict[ScriptIdT, typing.Tuple[str, str]] = {}
+
+
+def register_script(script: str) -> ScriptIdT:
+    global SCRIPTS
+    # NOTE(sileht): We don't use sha, in case of something server side change the script sha
+    script_id = ScriptIdT(uuid.uuid4())
+    SCRIPTS[script_id] = (
+        hashlib.sha1(script.encode("utf8")).hexdigest(),  # nosec
+        script,
+    )
+    return script_id
+
+
+async def run_script(
+    redis: utils.RedisStream,
+    script_id: ScriptIdT,
+    keys: typing.Tuple[str, ...],
+    args: typing.Optional[typing.Tuple[typing.Union[str], ...]] = None,
+) -> typing.Any:
+    global SCRIPTS
+    sha, script = SCRIPTS[script_id]
+    if args is None:
+        args = keys
+    else:
+        args = keys + args
+    try:
+        return await redis.evalsha(sha, len(keys), *args)
+    except aredis.exceptions.NoScriptError:
+        newsha = await redis.script_load(script)
+        if newsha != sha:
+            LOG.error(
+                "wrong redis script sha cached",
+                script_id=script_id,
+                sha=sha,
+                newsha=newsha,
+            )
+            SCRIPTS[script_id] = (newsha, script)
+        return await redis.evalsha(newsha, len(keys), *args)
+
+
+PUSH_PR_SCRIPT = register_script(
+    """
+local bucket_key = KEYS[1]
+local bucket_sources_key = KEYS[2]
+local scheduled_at_timestamp = ARGV[1]
+local source = ARGV[2]
+local score = ARGV[3]
+
+-- Add the source to the pull request sources
+redis.call("XADD", bucket_sources_key, "*", "source", source, "score", score)
+-- Add this pull request to the org bucket if not exists
+redis.call("ZADD", bucket_key, "NX", score, bucket_sources_key)
+-- Add the org bucket to the stream list
+redis.call("ZADD", "streams", "NX", scheduled_at_timestamp, bucket_key)
+"""
+)
```

Comment:
Sounds like this could just be a pipe, not a script? would be simpler

Reply:
I imagined computing the score in lua, redis-lua has msgpack built-in so we can read the streams
to compute score if needed.